### PR TITLE
Fix: mdsmisc.GetXYSignalXd set TimeContext based on SegLimit, isLong?

### PR DESCRIPTION
**Pull request by zack-vii**
It Dimensions are stored as int64 delta need to be converted to sec in order to get correct data.

### CODE REVIEW

**Comment by zack-vii** [comment]
retest this please

**Comment by GabrieleManduchi** [comment]
You are assuming that in case long is used, the time unit is nanosecond. This is not true, indeed we often use int64 to store UTC time expressed in milliseconds.

**Comment by GabrieleManduchi** [comment]
When we decided to use double representation, we got rid of the implicit conversion of all times in nanoseconds unit, and therefore now every time is a number related to the chosen units, so no division / multiplication for 1e9 must be done.

**Comment by zack-vii** [comment]
this is just because in xtreeshr we do the same. i would be happy to get rid of it in xtreeshr but this would break current behaviour. bottomline is we need to treat int64 in mdsmisc thesame way as we do in xtreeshr.

**Comment by GabrieleManduchi** [comment]
Ok, but what xtreeshr does is totally transparent outside XTreeGetTimedRecord.  
In ScopeUtilities.c,  estimateNumSamples does an estimation of the number of samples based on the actual delta value, that is totally unrelated to the way XTreeGetTimedRecord works. 
By the way, I remember to have  already removed that division by 1e9 some time ago since it broke jScope when zooming into waveformes with absolute UTC time.

**Comment by GabrieleManduchi** [comment]
We may later think about a different implementation of XTreeshr without the implicit conversion to nanoseconds, but this is unrelated to ScopeUtilities

**Comment by GabrieleManduchi** [comment]
...and if I am not wrong you already worked out the xtreeshr stuff.....

**Comment by GabrieleManduchi** [comment]
...ok, but without removing the 1e9 stuff....

**Comment by zack-vii** [comment]
hmm well the connection with xtree is duevto thebsettimecontext. and yes you are right i need to fix min and max too. the 1e9 in mdsmisc is only to compensate the 1e9 in xtreeshr. the output shoult then be in tge same data type and scale as the source.

**Comment by GabrieleManduchi** [comment]
But, again, you should disregard completely the 1e9 stuff outside the XTreeGetTimedRecord() internals.

**Comment by zack-vii** [comment]
min max is not affected as it will use NULL instead. 
again if mdsmisc want to make use of settimecontext we need to treat long differently. we can use a dtype_q delta or compensate the 1e9. as i said the output is not affected.

**Comment by GabrieleManduchi** [comment]
Maybe I am missing something.....why should I be concerned about 1E9  even if I use SetTimeContext passing int64 times? In normal use cases int64 is used to represent absolute times that may have ns units (EPICS times) or ms unit (UTC times). In any case the passed min and max to SetTimeContext will be consistent with the signals being read.  I don'care If they are internally multiplied /divided by 1E9 in xtreeshr, what I am interested in is that signals read afterwords (of course with times consistent to what has been passed to SetTimeContext) return only the portion within min and max.
In other words, no time unit is assumed by SetTimeContext and XTreeGetTimedRecord().

**Comment by zack-vii** [comment]
What you are missing is that the delta is calculated based on the segment Limits with would be a direct cast of int64 to double. Before mdsmisc put delta in settimecontext as is (direct cast as double not int64). xtreeshr well interpret the as-is delta in seconds. It will also notice the data's dim is in int64 and interpret it as nanoseconds. Consequently it will convert the one in the other and introduce the factor 1e9 to delta. To make it compare as is units with as is units we can compensate delta by dividing 1e9.

**Comment by GabrieleManduchi** [comment]
Ah, I see.... if I understand, you are in this way computing delta in double. In case segment limits were given in int64  (i.e.  jScope is dealing with absolute times) there is a misinterpretation by xtreeshr of the information stored is SetTimeContext. Why not converting also delta in int64 (from the computed double value) in this case (absolute times)  instead of exposing the internal xtreeshr 1E9 stuff in mdsmisc?

**Comment by zack-vii** [comment]
Basically we could do that but dividing delta by 1e9 is a one-liner while changing the datatype of delta would require some more elements. At minimum you would need to change deltaP to use DTYPE_Q and point to a int64.

**Comment by GabrieleManduchi** [comment]
I think that  changing  deltaP to use DTYPE_Q and point to a int64 when isLong is returned true by estimateNumSamples()  is much better, even if it requires some lines more.

In this case the xtreeshr internals are not exposed elsewhere and if we decide to change anything inside xtreeshr there are no side effects.

**Comment by zack-vii** [comment]
maybe, but if we change xtreeshr we can as well delete the line. in fact, we would want to get rid of the entire extra handling anyway. so we need to touch the file anyway.

**Comment by GabrieleManduchi** [comment]
"""""maybe, but if we change xtreeshr we can as well delete the line. in fact, we would want to get rid of the entire extra handling anyway. so we need to touch the fike anyway"""""

This is exactly what I want to avoid!!  In few months no one will remember this and in case xtreeshr will be changed sometimes in future,  jScope will break and it will take time to figure out what happened. Another thing is getting rid of the entire extra handling. If we'll do it after changing xtreeshr, good, we have cleaned code, if we forget or decide to postpone it, nothing gets broken anyway.

**Comment by zack-vii** [comment]
well that would be the changelog:
```c
-    if (isLong) delta /= 1e9; // quick compensation of xtreeshr conversion
-    deltaP = (delta>1e-9) ? &deltaD : NULL;
+    // since xtreeshr has a special handling for int64 timestamps we need to address this here
+    // one way would be to compensate the 1e9 factor or we pass delta as DTYPE_Q
+    if (isLong) {
+      if (delta >= 1.5) {
+        deltaP = &deltaD;
+        *(int64_t*)&delta = (int64_t)(delta+.5);
+        deltaD.dtype = DTYPE_Q;
+      } else deltaP = NULL;
+    } else
+      deltaP = (delta>1e-9) ? &deltaD : NULL;
```
i would rather add a test that will alarm us and add a comment to xtresshr that will remind us. but I think it is unlikely that we will change xtreeshr. It suites us well with combined dimensions in double sec and int64 ns plotted in on scope.

**Review by GabrieleManduchi — APPROVED** [review]
Ok, what is important is to make sure hat we don't forget it in future
